openmetadata-service/src/main/java/org/openmetadata/service/apps/bundles/changeEvent/generic/GenericPublisher.java

```diff
@@ -67,79 +69,89 @@ public GenericPublisher(
-                subscriptionDestination.getId(), webhook.getEndpoint());
-        LOG.warn(message);
-        setErrorStatus(attemptTime, 400, "UnknownHostException");
-      } else {
-        message =
-            CatalogExceptionMessage.eventPublisherFailedToPublish(WEBHOOK, event, ex.getMessage());
-        LOG.error(message);
-      }
-      throw new EventPublisherException(message, Pair.of(subscriptionDestination.getId(), event));
+      handleException(attemptTime, event, ex);
     }
   }
 
   @Override
   public void sendTestMessage() throws EventPublisherException {
     long attemptTime = System.currentTimeMillis();
     try {
-      // Post Message to default
-      String json =
-          "This is a test message from OpenMetadata to confirm your webhook destination is configured correctly.";
-      if (webhook.getEndpoint() != null) {
-        if (webhook.getSecretKey() != null && !webhook.getSecretKey().isEmpty()) {
-          String hmac =
-              "sha256="
-                  + CommonUtil.calculateHMAC(decryptWebhookSecretKey(webhook.getSecretKey()), json);
-          postWebhookMessage(this, getTarget().header(RestUtil.SIGNATURE_HEADER, hmac), json);
-        } else {
-          postWebhookMessage(this, getTarget(), json);
-        }
-      }
+      prepareAndSendMessage(TEST_MESSAGE_JSON, getTarget());
     } catch (Exception ex) {
-      Throwable cause = ex.getCause();
-      String message;
-      if (cause != null && cause.getClass() == UnknownHostException.class) {
-        message =
-            String.format(
-                "Unknown Host Exception for Generic Publisher : %s , WebhookEndpoint : %s",
-                subscriptionDestination.getId(), webhook.getEndpoint());
-        LOG.warn(message);
-        setErrorStatus(attemptTime, 400, "UnknownHostException");
-      } else {
-        message = CatalogExceptionMessage.eventPublisherFailedToPublish(WEBHOOK, ex.getMessage());
-        LOG.error(message);
+      handleException(attemptTime, ex);
+    }
+  }
+
+  private void prepareAndSendMessage(String json, Invocation.Builder target) {
+    if (!CommonUtil.nullOrEmpty(webhook.getEndpoint())) {
+
+      // Add HMAC signature header if secret key is present
+      if (!CommonUtil.nullOrEmpty(webhook.getSecretKey())) {
+        String hmac =
+            "sha256="
+                + CommonUtil.calculateHMAC(decryptWebhookSecretKey(webhook.getSecretKey()), json);
+        target.header(RestUtil.SIGNATURE_HEADER, hmac);
       }
-      throw new EventPublisherException(message);
+
+      // Add custom headers if they exist
+      Map<String, String> headers = webhook.getHeaders();
+      if (!CommonUtil.nullOrEmpty(headers)) {
+        headers.forEach(target::header);
+      }
+
+      Webhook.HttpMethod httpOperation = webhook.getHttpMethod();
+      postWebhookMessage(this, target, json, httpOperation);
+    }
+  }
+
+  private void handleException(long attemptTime, ChangeEvent event, Exception ex)
```

Comment:
why are there handleException , i see most of the code is same , can you remove the duplicate code

Reply:
Noted.